RegClient: implement Ping() method

edigaryev opened the pull request:
Problem: it's not possible to validate whether the credentials provided to the `RegClient` will work for a given repository without doing some operation on a manifest or a blob, which is not possible for CLI utilities using `github.com/regclient/regclient` providing a functionality similar to `docker login`, where we won't know the reference and a tag (or content digest) yet.

Solution: provide a `Ping()` method that will fail (see `end-1` [here](https://github.com/opencontainers/distribution-spec/blob/main/spec.md#endpoints)) when authentication is not valid or OCI Distribution Specification is not supported.

edigaryev commented:
@sudo-bmitch, would you mind taking a look at this?

sudo-bmitch commented:
I've been thinking this one over for a bit. It looks like it's following the `RepoList` implementation, which is a bit of a sore spot that I'd like to eventually deprecate/replace. Knowing that OCI also wants to do some work in that space has me holding off on fixing that API right now.

Part of fixing that may involve expanding `Ref` to support partial content, e.g. registry, or registry+repo, but not a full reference to a manifest. Then a `Ref` could be passed to `RepoList` rather than a hostname string. And I think it could be expanded to support ocidir schemes to find directories with an `oci-layout` and `index.json` file.

For the `Ping`, I'm still deciding between a few options:

1. Accept as is. It fills a need and isn't significant to maintain.
2. Rework to follow the goals with `RepoList`, and consider what support for `ocidir` may look like.
3. Focus on the end use case, validating credentials, and do something closer to the auth layer.
4. Push back, since a valid auth doesn't indicate if an action would be allowed, and recommend pushing an small blob or pulling a small manifest instead.

Since auth is currently internal, I'm leaning against 3. And there are a lot of tools providing this feature, so I'm also leaning against 4. It's mostly a question of whether to accept this and deprecate later for a different implementation, or rework things now for `Ref` and make this the first use case.

edigaryev commented:
If I understood correctly, you want to support `Ref` in `Ping()` instead of `hostname`, but for `ocidir` it'll always be a stub code path that returns no error, right?

And based on your analysis, the only difference in the remaining two options (1 and 2) is that two methods will need to be deprecated with option 1, compared to one method with option 2, which doesn't seem to be a big deal.

On the other hand, going with option 1 has a pros of capturing the user-base now: the lack of `Ping()` is the only thing that prevents us from moving forward with this package currently.

sudo-bmitch commented:
I've got `Ref` refactored now which should get this on a better path. For ocidir, I think checking for the existence and readability of the path could be useful. The `Ping` is not only checking user credentials, but also accessibility of the remote server, so accessibility of the local directory feels like a good equivalent.

This can also use some tests and I'd like to get it connected into the `regctl registry login` and `regctl registry set` command to at least warn if the server cannot be reached using the requested login and configuration. I'm happy to do a joint PR on that if you'd like. I should have time on Wednesday to knock that out.

edigaryev commented:
> I've got `Ref` refactored now which should get this on a better path.

Great news! I've rebased this PR and switched the `Ping()` to use the `ref.Ref`. Some further questions:

1. Do you want to also support `Ping()` for `ocidir` in this PR?
2. Do you want to also switch `RepoList()` to `ref.Ref` in this PR?

> This can also use some tests

What do you think about Golang-based tests (contrary to the `build/ci-test.sh`) using `github.com/testcontainers/testcontainers-go` and running `registry:2`?

sudo-bmitch commented:
> > I've got `Ref` refactored now which should get this on a better path.
> 
> Great news! I've rebased this PR and switched the `Ping()` to use the `ref.Ref`. Some further questions:
> 
>     1. Do you want to also support `Ping()` for `ocidir` in this PR?

I do, along with some other refactoring (I'm looking at updating the interface and including a struct in the return in case someone wants headers or other data from the ping).

>     2. Do you want to also switch `RepoList()` to `ref.Ref` in this PR?

I'm going to give that one some time. Users are dependent on it (part of the reason to not rush this PR), and the wg-artifacts group may give some direction on what OCI should be implementing.

> > This can also use some tests
> 
> What do you think about Golang-based tests (contrary to the `build/ci-test.sh`) using `github.com/testcontainers/testcontainers-go` and running `registry:2`?

I wouldn't touch those tests for this. They're mostly a compatibility smoke test designed to replicate end user experiences. I may add more registries and a few more commands to the list in the future, but for the most part I ignore those when looking at coverage reports.

The majority of the tests are Go unit tests. There's a need to refactor those unit tests a bit to consolidate the registry mocking, but I'd like to get change merged first. I'll take a shot at it this afternoon. Given all the refactoring I'm thinking of, I'll probably end up pushing it as a separate PR (too much to fit in a review), but keeping your commits as the starting point to be sure your name is listed in the release as a contributor.

sudo-bmitch commented:
@edigaryev when you have a chance, have a look at #590 and let me know it that still works for your use case.

edigaryev commented:
Closing in favor of #590.
